### ISSUE TITLE
vertice and texture cooridnates normalized mode

### ISSUE THREAD

**Issue by grajen3** [OneB1t/VcMOSTRenderMqb#2]
I was playing with VNC client with some generated images being served and noticed that center of my generated content is not in the center of VC (and overall some blurriness)

![image](https://github.com/user-attachments/assets/6e0d11f3-5023-461a-b5e4-95f4aef42335)
Is my current stuff - but that green box with navigation cue had to be rendered off center by trial and error (and even now it doesn't feel centered hehe). Worse part is that "your location" on the map is not centered - it is a bit to the right off center, but this part I can't control myself unless I add custom clipping after the fact.

I suspect those to be "at fault":
https://github.com/OneB1t/VcMOSTRenderMqb/blob/4f726cf6ac5b53074988a704ac04dbf8fe89968a/opengl-render-qnx/opengl.cc#L88-L113
I assume this is some clipping that attempts to remove some "junk" from screen recordings tailored to nav app(s) you use?

Would you accept a change that allow to use fully normalized coordinates? so vertices would be from -1 to 1 and tex coordinates from 0 to 1 and that could be toggle with CLI switch like `--no-clipping` or something along the lines (as to not change default behavior)?

Unless my understanding is wrong and this is not just for screenrecording, but also for VC display mapping?

**Comment by OneB1t** [comment]
Yes this is there to clip unused parts of image. I usually run windows version where you can see perfectly what will be rendered. To get rid of this behaviour just use

GLfloat landscapeVertices[] = { 
    -1.0f,  1.0f, 0.0f,  // Top Left 
     1.0f,  1.0f, 0.0f,  // Top Right 
     1.0f, -1.0f, 0.0f,  // Bottom Right 
    -1.0f, -1.0f, 0.0f   // Bottom Left 
 }; 

I will add no-clip option for next releases

**Comment by grajen3** [comment]
> I usually run windows version where you can see perfectly what will be rendered.

Yeah, did that today (had to setup visual studio heh) and will be using that instead of vnc client I used for testing before :)

Not sure if it's just me but windows app seems unresponsive? it does work and renders stuff, I just can't move window or stuff like that

**Comment by OneB1t** [comment]
![image](https://github.com/user-attachments/assets/7f1356be-9fbc-465d-8ef7-84f1f3aa7788)

yes the application is somehow unresponsive no clue why i never cared to fix it it probably requires some sleep in order to deal with windowing system events

**Comment by OneB1t** [comment]
unresponsive window fixed in https://github.com/OneB1t/VcMOSTRenderMqb/commit/4546d582d7040fdf73ae75fa32686c0e2acecf59

**Comment by OneB1t** [comment]
@grajen3 this "Oil" temp is nice can you also try to add battery voltage? 😈

**Comment by grajen3** [comment]
Oh the oil stuff from my screenshot is all inside LSD - reading it there and using BAP stuff to set it as one of navigation related data fields, but I will actually be dropping that if the google map rendering prove to be working fine ... some day :) )

I'm not doing any rendering of the google map on the car's hw (other than relying on your VNC client) and all the rendering is on the phone right now, so I don't even know how would I use those persistance channels ( I saw https://github.com/OneB1t/VcMOSTRenderMqb/issues/1 about it, but this seems like something accessible from car itself that could be exposed with some custom server if needed)

I was thinking about replacing what I have now with BAP and just render oil (and whatever else) on the phone somewhere on top of the map and get that data from exlap (using same thing as https://github.com/jilleb/mqb-pm so there is something for android already built that makes reading from exlap easier and not something that has to be done from scratch) - oil temp is for sure available from exlap and I'm pretty sure battery voltage is as well ( this has reference to it https://github.com/jilleb/mqb-pm/blob/c47ca0ed3e277ac72fa525803f6c9c5b3aaeed5b/app/src/main/java/com/mqbcoding/stats/DashboardFragment.java#L2691-L2696 )

**Comment by OneB1t** [comment]
There is a way to draw over map image but my font is quite ugly one :-/ maybe later i will invest more time into nicer font. Car itself is serving exlap data over port 25010 so it is possible to connect there over WIFI and read car data from that endpoint.

**Comment by grajen3** [comment]
I would just like to do all the rendering in single place - given that I have map and I already custom render navigation cue (green box stuff that mimics google maps style) it just make sense to me to render rest of the things on the phone too (if data is available and it is).

Also think the API for drawing is a bit nicer with Android than dealing with OpenGL directly - so it's just "easier" to work with

**Comment by OneB1t** [comment]
on android it will be light years better for font rendering and also for exlap client (as there are libraries which can be used). Exlap also expose consumption data, dsg oil data, performance data (on cars which have that). etc..

**Comment by grajen3** [comment]
> There is small python software on my repo developed by @jilleb https://github.com/OneB1t/VcMOSTRenderMqb/blob/main/RESEARCH/exlapResearch/exlap_client_v2.py which can be easily transformed into java app (chatGPT)

I think I'll just copy most of https://github.com/martoreto/aauto-vex-vag/tree/master/app/src/main/java/com/github/martoreto/aauto/vex/vag and just adjust transport stuff (this is using android auto custom vehicle extensions transport AFAIK which would require android auto connection to work, but nothing in my custom renderer rely on actually runing android auto so it would be pointless limitation if I can connect to it over 25010 port)

**Comment by grajen3** [comment]
Slowly getting there :)
![image](https://github.com/user-attachments/assets/0b5e0729-b458-48c7-a76a-f12f399005fb)

**Comment by grajen3** [comment]
And battery voltage reading and rendering on the phone working:
![image](https://github.com/user-attachments/assets/7000ef08-6f21-4dcb-9939-ca4d2861b3ec)

Should print oil temp next to battery voltage, but so far I was testing without even starting engine - I'd need warm engine to verify because we don't get any values until it's 50 deg - will check on my next trip.

I'll try clean up my code a bit and share the project, but the part that syncs google maps navigation with this custom renderer is pretty wonky so probably will omit that at the start at least (I'm using additional server for that which make running it not the greatest for other possibly interested in playing with it and would need to redo some things to avoid need for it)

**Comment by OneB1t** [comment]
As you are progressing quite nicely, I think I will focus my efforts on seeing if it is possible to increase the framerate of the video stream from MIB to VC. It is currently limited to 10 fps, but I hope there might be some flexibility.

The exlap client in plain C was tough, and I had been waiting to tackle it during the long winter nights. However, if there is a way to program everything on top of the phone, that would be a much easier and faster approach."

**Comment by OneB1t** [comment]
New version have support for config.txt with custom coordinates. It must be in the same folder as binary.

**Comment by grajen3** [comment]
Awesome! Had a bit of a break from playing with VC stuff, but given that there is config for those vertices now - I'll close this issue as it is now done :)

**Comment by OneB1t** [comment]
Waiting for you to join the party 👍 I think i also found a way to increase map FPS with lsd.jar patch


de\vw\mib\asl\internal\mostkombi\streamsink\usecases\ChangeDataRateSequence.java
```
private void _stepChangeDataRate() {
        switch (this.getStreamSinkDataRate()) {
            case 0: {
                this.getDisplayManagementService().setDataFrameRate(0); // invisible
                break;
            }
            case 1: {
                this.getDisplayManagementService().setDataFrameRate(1); // invisible
                break;
            }
            case 2: {
                **this.getDisplayManagementService().setDataFrameRate(10); // visible**
                break;
            }
        }
    }
```

As you have more knowledge how to not brick unit while messing with lsd.jar maybe you can try?

**Comment by pieh** [comment]
> [@grajen3](https://github.com/grajen3): Waiting for you to join the party 👍 I think i also found a way to increase map FPS with lsd.jar patch
> 
> de\vw\mib\asl\internal\mostkombi\streamsink\usecases\ChangeDataRateSequence.java
> 
> ```
> private void _stepChangeDataRate() {
>         switch (this.getStreamSinkDataRate()) {
>             case 0: {
>                 this.getDisplayManagementService().setDataFrameRate(0); // invisible
>                 break;
>             }
>             case 1: {
>                 this.getDisplayManagementService().setDataFrameRate(1); // invisible
>                 break;
>             }
>             case 2: {
>                 **this.getDisplayManagementService().setDataFrameRate(10); // visible**
>                 break;
>             }
>         }
>     }
> ```
> 
> As you have more knowledge how to not brick unit while messing with lsd.jar maybe you can try?
> 
> this.getDisplayManagementService().setDataFrameRate(20); or this.getDisplayManagementService().setDataFrameRate(30);

Oh, I've missed this comment before - I'll try to remember checking this out next time I find time ... and willingness :) to play with this stuff.

I'm not sure it will be useful for me, as I already attempt to limit frame rate to 10 in my custom stuff, because sometimes I will get phone overheating messages and android just kind of soft-killing my app, so trying to get more framerate for this use case (probably more about my not very efficient implementation hehe) doesn't seem that useful to my usage (the ~10 fps seems mostly good enough for me - it's not super responsive and it also seems up to ~1 sec lagging behind that I can see on crossings - like I already made the turn, but map in VC shows state still before that turn etc - this all might be my implementation fault, but for what I use it for it's good enough).

I'll try checking messing with that regardless (if I don't forget) socommunity at least know how far things could be pushed this way

**Comment by grajen3** [comment]
> [@grajen3](https://github.com/grajen3): Waiting for you to join the party 👍 I think i also found a way to increase map FPS with lsd.jar patch
> 
> de\vw\mib\asl\internal\mostkombi\streamsink\usecases\ChangeDataRateSequence.java
> 
> ```
> private void _stepChangeDataRate() {
>         switch (this.getStreamSinkDataRate()) {
>             case 0: {
>                 this.getDisplayManagementService().setDataFrameRate(0); // invisible
>                 break;
>             }
>             case 1: {
>                 this.getDisplayManagementService().setDataFrameRate(1); // invisible
>                 break;
>             }
>             case 2: {
>                 **this.getDisplayManagementService().setDataFrameRate(10); // visible**
>                 break;
>             }
>         }
>     }
> ```
> 
> As you have more knowledge how to not brick unit while messing with lsd.jar maybe you can try?
> 
> this.getDisplayManagementService().setDataFrameRate(20); or this.getDisplayManagementService().setDataFrameRate(30);

I just noticed this - ops pretty late :)

I'll try to check if this indeed can increase frame rate - just probably wouldn't use this for my usage even if it could, because I already do attempt limit frame rate in my custom app, because phone sometimes do get "phone is overheating" notifications and it somewhat kills/suspend my app then, so instead of unlimited FPS - I try to fit ~10 FPS already.

Probably will have to use regular VNC screen sharing app and try to play some ~30FPS video (ideally with numbers counting for each frame :D) - I do remember you were showing either video or description of doing something similar - just rendering with opengl to try to find out actual possible frame rate before (and just hope I won't have to implement my version of something like that - it's already difficult to find time and motivation nowadays for me to keep tinkering heh)

**Comment by OneB1t** [comment]
@grajen3
yes it works :-) i was able to go up to 30fps. Problem is now with frame delivery from phone as my implementation is just dump zlib without any tricks and its cpu limited

If i reimplement VNC client with tightVNC encoding i think that it will be possible to decrease cpu load
other approach is to buy faster phone
